src/main/java/hudson/remoting/Engine.java

```diff
@@ -678,6 +643,23 @@ public void afterChannel(@Nonnull JnlpConnectionState event) {
         }
     }
 
+    private JnlpEndpointResolver createEndpointResolver(List<String> jenkinsUrls) {
+        JnlpEndpointResolver resolver;
+        if (instanceIdentity == null) {
```

Comment:
Wouldn't it be better readable and more intuitive if `directUrl` would be used to determine the behaviour?
```suggestion
        if (isEmpty(directUrl)) {
```